src/analyses/lexical_loops.h

```diff
@@ -0,0 +1,175 @@
+/*******************************************************************\
+
+Module: Compute lexical loops in a goto_function
+
+Author: Diffblue Ltd
+
+\*******************************************************************/
+
+/// \file
+/// Compute lexical loops in a goto_function.
+///
+/// A lexical loop is a block of (lexically) contiguous GOTO program
+/// instructions with a single entry edge at the top and a single backedge
+/// leading from bottom to top. This is a subset of the natural loops, which is
+/// cheaper to compute and spots most natural loops generated from typical C
+/// code.
+
+#ifndef CPROVER_ANALYSES_LEXICAL_LOOPS_H
+#define CPROVER_ANALYSES_LEXICAL_LOOPS_H
+
+#include <iosfwd>
+#include <set>
+#include <stack>
+
+#include <goto-programs/goto_model.h>
+
+#include "loop_analysis.h"
+
+/// Main driver for working out if a class (normally goto_programt) has any
+/// lexical loops. \ref compute takes an entire goto_programt, iterates over the
+/// instructions and for any backwards jumps attempts to find out if it's a
+/// lexical loop.
+///
+/// All instructions in a lexical loop are stored into \ref loop_map, keyed by
+/// their head - the target of the backwards goto jump.
+///
+/// \tparam P: the program representation and needs:
+///   * [field] instruction which is an iterable of type T.
+/// \tparam T: iterator of the particular node type, e.g.
+///   std::list<...>::iterator. The object this iterator holds needs:
+///   * [function] is_backwards_goto() returning a bool.
+///   * [function] get_target() which returns an object that needs:
+///     * [field] location_number which is an unsigned int.
+template <class P, class T>
+class lexical_loops_templatet : public loop_analysist<P, T>
+{
+  typedef loop_analysist<P, T> parentt;
+
+public:
+  typedef typename parentt::loopt lexical_loopt;
+
+  void operator()(P &program)
+  {
+    compute(program);
+  }
+
+  lexical_loops_templatet() = default;
+
+  explicit lexical_loops_templatet(P &program)
+  {
+    compute(program);
+  }
+
+  bool all_cycles_in_lexical_loop_form() const
```

Comment:
This also doesn't seem to be used anywhere in this PR. Do you have a usage in mind for this?

Reply:
Yes, when we use it in symex we'll want to check if the function contains cycles that aren't lexical loops